task/prepare.go

```diff
@@ -0,0 +1,80 @@
+package task
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/golang/glog"
+	livepeerAPI "github.com/livepeer/go-api-client"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/stream-tester/model"
+	"github.com/livepeer/stream-tester/segmenter"
+)
+
+func RecordStream(ctx context.Context, lapi *livepeerAPI.Client, playbackID string, osSess drivers.OSSession) (string, error) {
+	glog.Infof("made it into record function")
+	fullPath := videoFileName(playbackID)
+	glog.Infof("full path %s", fullPath)
+	fir, err := osSess.ReadData(ctx, fullPath)
+	glog.Infof("inputOS %b", osSess.IsExternal())
+	if err != nil {
+		return "", fmt.Errorf("error reading data from source OS url=%s err=%w", fullPath, err)
+	}
+	sourceFile, err := readFile(fir)
+	fir.Body.Close()
+	if err != nil {
+		return "", err
+	}
+	defer sourceFile.Close()
+	var sourceFileSize int64
+	if fir.Size != nil {
+		sourceFileSize = *fir.Size
+	}
+
+	streamName := fmt.Sprintf("vod_recording_%s", time.Now().Format("2006-01-02T15:04:05Z07:00"))
+	stream, err := lapi.CreateStreamEx(streamName, true, nil)
+	if err != nil {
+		return "", nil
+	}
+	defer lapi.DeleteStream(stream.ID)
+
+	gctx, gcancel := context.WithCancel(ctx)
+	defer gcancel()
+	segmentsIn := make(chan *model.HlsSegment)
+	if err = segmenter.StartSegmentingR(gctx, sourceFile, true, 0, 0, segLen, false, segmentsIn); err != nil {
+		return "", err
+	}
+	ws := fileWriter(sourceFileSize)
+	defer ws.Close()
+	var transcoded [][]byte
+	err = nil
+
+	for seg := range segmentsIn {
+		if seg.Err == io.EOF {
+			break
+		}
+		if seg.Err != nil {
+			err = seg.Err
+			glog.Errorf("Error while segmenting playbackID=%s err=%v", playbackID, err)
+			break
+		}
+		glog.V(model.VERBOSE).Infof("Got segment seqNo=%d pts=%s dur=%s data len bytes=%d\n", seg.SeqNo, seg.Pts, seg.Duration, len(seg.Data))
+		started := time.Now()
+		_, err = lapi.PushSegment(stream.ID, seg.SeqNo, seg.Duration, seg.Data)
+		if err != nil {
+			glog.Errorf("Segment push playbackID=%s err=%v\n", playbackID, err)
+			break
+		}
+		glog.V(model.VERBOSE).Infof("Transcode %d took %s\n", len(transcoded), time.Since(started))
+	}
+	if err == io.EOF {
+		err = nil
+	}
+	if err != nil {
+		return "", err
+	}
+
```

Comment:
I think it would be nice to check if the recording is correct here. We can start by just checking if it has the right duration. If it is not, we should make sure to fail the task instead of creating the asset with broken playback.

Reply:
Good point, will add that

Reply:
Can we defer this check to a fast follow-up PR to get HLS playback released sooner? May not be trivial to perform this check unsure how long implementation will take.

Reply:
Sure